File: hs/src/Day10.hs
-- |
-- Module:         Day10
-- Description:    <https://adventofcode.com/2024/day/10 Day 10: Hoof It>
module Day10 (part1, part2) where

import Data.Char (digitToInt, isDigit)
import Data.IntMap qualified as IntMap (findWithDefault, fromListWith)
import Data.List.NonEmpty (NonEmpty ((:|)))
import Data.Map (Map)
import Data.Map qualified as Map (elems, fromListWith, fromSet, toList)
import Data.Monoid (Sum (Sum, getSum))
import Data.Set (Set)
import Data.Set qualified as Set (empty, member, singleton, size)
import Data.Text (Text)
import Data.Text qualified as T (lines, unpack)

parse :: Text -> NonEmpty (Set (Int, Int))
parse input = IntMap.findWithDefault Set.empty `flip` elevations <$> 0 :| [1 .. 9]
  where
    elevations =
      IntMap.fromListWith (<>) $
        [ (digitToInt c, Set.singleton (y, x))
        | (y, line) <- zip [0 ..] $ T.lines input,
          (x, c) <- zip [0 ..] $ T.unpack line,
          isDigit c
        ]

adj :: (Int, Int) -> [(Int, Int)]
adj (y, x) = [(y - 1, x), (y, x - 1), (y, x + 1), (y + 1, x)]

bfs :: (Semigroup a) => ((Int, Int) -> a) -> NonEmpty (Set (Int, Int)) -> Map (Int, Int) a
bfs start (zero :| elevations) = foldl bfs' (Map.fromSet start zero) elevations
  where
    bfs' acc points =
      Map.fromListWith (<>) $
        [ (q, m)
        | (p, m) <- Map.toList acc,
          q <- adj p,
          q `Set.member` points
        ]

part1 :: Text -> Int
part1 input = sum $ Set.size <$> bfs Set.singleton (parse input)

part2 :: Text -> Int
part2 input = getSum $ mconcat $ Map.elems $ bfs (const $ Sum 1) (parse input)

File: py/aoc2024/day10.py
"""
Day 10: Hoof It
"""

from operator import add, or_
from typing import Callable

SAMPLE_INPUT = """
89010123
78121874
87430965
96549874
45678903
32019012
01329801
10456732
"""


def _parse(data: str) -> list[set[tuple[int, int]]]:
    elevations = [set() for _ in range(10)]
    for y, line in enumerate(data.splitlines()):
        for x, char in enumerate(line):
            if "0" <= char <= "9":
                elevations[int(char)].add((y, x))
    return elevations


def _adj(point: tuple[int, int]) -> list[tuple[int, int]]:
    y, x = point
    return [(y - 1, x), (y, x - 1), (y, x + 1), (y + 1, x)]


def _bfs[T](
    elevations: list[set[tuple[int, int]]],
    start: Callable[[tuple[int, int]], T],
    plus: Callable[[T, T], T],
):
    acc = {key: start(key) for key in elevations[0]}
    for points in elevations[1:]:
        tmp = {}
        for key, value in acc.items():
            for point in _adj(key):
                if point in points:
                    tmp[point] = plus(value, tmp[point]) if point in tmp else value
        acc = tmp
    return acc


def part1(data: str) -> int:
    """
    >>> part1(SAMPLE_INPUT)
    36
    """
    return sum(
        len(value) for value in _bfs(_parse(data), lambda point: {point}, or_).values()
    )


def part2(data: str) -> int:
    """
    >>> part2(SAMPLE_INPUT)
    81
    """
    return sum(_bfs(_parse(data), lambda _: 1, add).values())


parts = (part1, part2)
